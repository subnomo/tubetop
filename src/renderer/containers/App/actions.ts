import { SongData } from '../../components/Song';
import {
  PLAY_SONG,
  ADD_SONG,
  ADD_SONGS,
  EDIT_SONG,
  EDIT_SONGS,
  REMOVE_SONG,
  CLEAR_SONGS,
} from './constants';

export interface AppAction {
  type: string;
  song?: SongData;
  songs?: SongData[];
  index?: number;
}

export function playSong(index: number): AppAction {
  return {
    type: PLAY_SONG,
    index,
  };
}

export function addSong(song: SongData): AppAction {
  return {
    type: ADD_SONG,
    song,
  };
}

export function addSongs(songs: SongData[]): AppAction {
  return {
    type: ADD_SONGS,
    songs,
  };
}

export function editSong(song: SongData): AppAction {
  return {
    type: EDIT_SONG,
    song,
  };
}

export function editSongs(songs: SongData[]): AppAction {
  return {
    type: EDIT_SONGS,
    songs,
  };
}

export function removeSong(index: number): AppAction {
  return {
    type: REMOVE_SONG,
    index,
  };
}

export function clearSongs(): AppAction {
  return {
    type: CLEAR_SONGS,
  };
}
