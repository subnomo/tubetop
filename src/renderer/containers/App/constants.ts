export const ADD_SONG = 'tubetop/App/ADD_SONG';
export const REMOVE_SONG = 'tubetop/App/REMOVE_SONG';
export const PLAY_SONG = 'tubetop/App/PLAY_SONG';
