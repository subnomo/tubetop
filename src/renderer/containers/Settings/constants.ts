export const SAVE_SETTINGS = 'tubetop/Settings/SAVE_SETTINGS';
