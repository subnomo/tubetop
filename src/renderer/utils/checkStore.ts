import { conformsTo, isFunction, isObject } from 'lodash';
import * as invariant from 'invariant';

/**
 * Validate the shape of redux store
 */
export default function checkStore(store: any) {
  const shape = {
    dispatch: isFunction,
    subscribe: isFunction,
    getState: isFunction,
    replaceReducer: isFunction,
    runSaga: isFunction,
    injectedReducers: isObject,
    injectedSagas: isObject,
  };

  invariant(
    conformsTo(store, shape),
    '(renderer/utils...) injectors: Expected a valid redux store',
  );
}
