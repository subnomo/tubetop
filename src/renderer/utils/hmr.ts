interface IHMROptions {
  type: 'js' | 'css';
  path: string;
  content: string;
}

const customizedHMRPlugin = {
  hmrUpdate: (opts: IHMROptions) => {
    const shouldUpdate = (path: string) => {
      const ignoredFiles = [
        /store/,
        /reducers/,
      ];

      for (let i = 0; i < ignoredFiles.length; i++) {
        if (ignoredFiles[i].test(path)) {
          return false;
        }
      }

      return true;
    };

    if (!shouldUpdate(opts.path)) {
      window.location.reload();
    }

    return true;
  }
};
