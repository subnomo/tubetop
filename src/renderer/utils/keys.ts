export const YOUTUBE_API_KEY = 'AIzaSyDwXKHjw1IoqsLQNf4yykNj6YEn-VlsOa8';
