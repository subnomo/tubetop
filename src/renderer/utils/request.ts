import 'whatwg-fetch';

/**
 * Parses the JSON returned by a network request
 *
 * @param  {object} response A response from a network request
 *
 * @return {object}          The parsed JSON from the request
 */
function parseJSON(response: Response): Promise<any> {
  if (response.status === 204 || response.status === 205) {
    return null;
  }

  return response.json();
}

/**
 * Checks if a network request came back fine, and throws an error if not
 *
 * @param  {object} response   A response from a network request
 *
 * @return {object|undefined} Returns either the response, or throws an error
 */
function checkStatus(response: Response): Response {
  if (response.status >= 200 && response.status < 300) {
    return response;
  }

  const error: any = new Error(response.statusText);
  error.response = response;
  throw error;
}

/**
 * Requests a URL, returning a promise
 *
 * @param  {string} url       The URL we want to request
 * @param  {object} [options] The options we want to pass to "fetch"
 *
 * @return {object}           The response data
 */
export default function request(url: string, options?: any): Promise<any> {
  return fetch(url, options)
    .then(checkStatus)
    .then(parseJSON);
}
