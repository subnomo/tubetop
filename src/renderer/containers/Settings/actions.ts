import { SAVE_SETTINGS } from './constants';

export function saveSettings(settings: any) {
  return {
    type: SAVE_SETTINGS,
    settings,
  };
}
