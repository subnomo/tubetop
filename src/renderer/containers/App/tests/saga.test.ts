import { put, takeLatest, SelectEffect, CallEffect, PutEffect } from 'redux-saga/effects';
import { cloneableGenerator, SagaIteratorClone } from 'redux-saga/utils';

import { AppAction, searchSuccess, searchFailure } from '../actions';
import { SEARCH_SONG } from '../constants';
import rootSaga, { isYoutubeLink, isYoutubePlaylist, search, searchExtra } from '../saga';

describe('isYoutubeLink', () => {
  it('should check if the given url matches a YouTube video', () => {
    expect(isYoutubeLink('https://www.youtube.com')).toBe(false);
    expect(isYoutubeLink('https://www.youtube.com/watch?v=dQw4w9WgXcQ')).toBe(true);
    expect(isYoutubeLink('https://www.youtube.com/playlist?list=PLBCF2DAC6FFB574DE')).toBe(false);
  });
});

describe('isYoutubePlaylist', () => {
  it('should check if the given url matches a YouTube playlist', () => {
    expect(isYoutubePlaylist('https://www.youtube.com')).toBe(false);
    expect(isYoutubePlaylist('https://www.youtube.com/watch?v=dQw4w9WgXcQ')).toBe(false);
    expect(isYoutubePlaylist('https://www.youtube.com/playlist?list=PLBCF2DAC6FFB574DE'))
      .toBe(true);
  });
});

const api_key = 'API_KEY';

describe('search Saga', () => {
  it('should return the results of the YouTube search', () => {
    const action = {
      type: SEARCH_SONG,
      query: 'abc',
    };

    const searchGenerator = cloneableGenerator(search)(action);

    function test() {
      const clone = searchGenerator.clone();

      const selectDescriptor = clone.next().value;
      expect(selectDescriptor).toMatchSnapshot();

      const callDescriptor = clone.next(api_key).value;
      expect(callDescriptor).toMatchSnapshot();

      // Returns
      const res = {
        items: [
          {
            snippet: {
              title: 'one',
            },
          },

          {
            snippet: {
              title: 'two',
            },
          },

          {
            snippet: {
              title: 'Deleted video',
            },
          },
        ],
      };

      const returnVal = clone.next(res).value;
      expect(returnVal).toEqual(res.items.slice(0, res.items.length - 1));
    }

    test();
    action.query = 'https://www.youtube.com/watch?v=dQw4w9WgXcQ';
    test();
    action.query = 'https://www.youtube.com/playlist?list=PLBCF2DAC6FFB574DE';
    test();
  });
});

describe('searchExtra Saga', () => {
  let searchExtraGenerator: SagaIteratorClone;

  beforeEach(() => {
    searchExtraGenerator = cloneableGenerator(searchExtra)({
      type: SEARCH_SONG,
      query: 'abc',
    });

    const selectDescriptor = searchExtraGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = searchExtraGenerator.next(api_key).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the searchSuccess action if it requests the data successfully', () => {
    const items: any[] = [
      {
        id: {
          kind: 'youtube#video',
        },
      },

      {
        kind: 'youtube#video',
      },

      {
        kind: 'youtube#playlistItem',
        snippet: {
          resourceId: {
            videoId: 'abc',
          },
        },
      },

      {},
    ];

    const callDescriptor2 = searchExtraGenerator.next(items).value;
    expect(callDescriptor2).toMatchSnapshot();

    const putDescriptor = searchExtraGenerator.next({ items }).value;
    expect(putDescriptor).toEqual(put(searchSuccess(items)));
  });

  it('should dispatch the searchFailure action if a response errors', () => {
    const items = [];
    const clone = searchExtraGenerator.clone();

    const response = new Error('Some error');
    const putDescriptor = clone.throw(response).value;
    expect(putDescriptor).toEqual(put(searchFailure(response)));

    // Returns
    clone.next();

    const callDescriptor2 = searchExtraGenerator.next(items).value;
    expect(callDescriptor2).toMatchSnapshot();

    const putDescriptor2 = searchExtraGenerator.throw(response).value;
    expect(putDescriptor2).toEqual(put(searchFailure(response)));

    // Returns
    searchExtraGenerator.next();
  });
});

describe('rootSaga Saga', () => {
  const rootSagaGenerator = rootSaga();

  it('should start task to watch for SEARCH_SONG action', () => {
    const takeLatestDescriptor = rootSagaGenerator.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(SEARCH_SONG, searchExtra));
  });
});
