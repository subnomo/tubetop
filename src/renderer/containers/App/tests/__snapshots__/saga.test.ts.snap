// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search Saga should return the results of the YouTube search 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`search Saga should return the results of the YouTube search 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://www.googleapis.com/youtube/v3/search?part=snippet&key=API_KEY&q=abc&maxResults=10",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`search Saga should return the results of the YouTube search 3`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`search Saga should return the results of the YouTube search 4`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://www.googleapis.com/youtube/v3/videos?part=snippet&key=API_KEY&id=dQw4w9WgXcQ",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`search Saga should return the results of the YouTube search 5`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`search Saga should return the results of the YouTube search 6`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://www.googleapis.com/youtube/v3/playlistItems?part=snippet&key=API_KEY&maxResults=50&playlistId=PLBCF2DAC6FFB574DE",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`searchExtra Saga should dispatch the searchFailure action if a response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`searchExtra Saga should dispatch the searchFailure action if a response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      Object {
        "query": "abc",
        "type": "tubetop/App/SEARCH_SONG",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`searchExtra Saga should dispatch the searchFailure action if a response errors 3`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://www.googleapis.com/youtube/v3/videos?part=contentDetails&key=API_KEY&id=",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`searchExtra Saga should dispatch the searchSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`searchExtra Saga should dispatch the searchSuccess action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      Object {
        "query": "abc",
        "type": "tubetop/App/SEARCH_SONG",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`searchExtra Saga should dispatch the searchSuccess action if it requests the data successfully 3`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://www.googleapis.com/youtube/v3/videos?part=contentDetails&key=API_KEY&id=%2C%2Cabc",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
